plugin/client_bots.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package plugin
+
+import (
+	"time"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/pkg/errors"
+)
+
+// EnsureBot ether returns an existing bot user or creates a bot user with
+// the specifications of the passed bot.
+// Returns the id of the bot created or existing.
+func (p *MattermostPlugin) EnsureBot(bot *model.Bot) (retBotId string, retErr error) {
+	// Must provide a bot with a username
+	if bot == nil || len(bot.Username) < 1 {
+		return "", errors.New("EnsureBot was passed a bad bot")
+	}
+
+	// If we fail for any reason, this could be a race between creation of bot and
+	// retreval from anouther EnsureBot. Just try the basic retrieve existing again.
+	defer func() {
+		if retBotId == "" || retErr != nil {
+			time.Sleep(time.Second)
```

Comment:
A second seems like a long time, can this be a progressive say, immediately, 128ms, 256ms, up to 3 times? Maybe we make this into a function that we can use in all kinds of `EnsureXXX` in plugins? Seems like a useful method to have.

Reply:
Makes sense but I created a separate ticket to add that: https://mattermost.atlassian.net/browse/MM-15267